src/lem/constrainer.rs

```diff
@@ -81,145 +117,320 @@ impl LEM {
         Ok(())
     }
 
-    fn on_concrete_path(concrete_path: &Option<Boolean>) -> Result<bool> {
-        match concrete_path {
-            None => Ok(true),
-            Some(concrete_path) => concrete_path
-                .get_value()
-                .ok_or_else(|| anyhow!("Couldn't check whether we're on a concrete path")),
-        }
-    }
-
     fn allocate_and_inputize_input<'a, F: LurkField, CS: ConstraintSystem<F>>(
         &'a self,
         cs: &mut CS,
         store: &mut Store<F>,
-        witness: &Witness<F>,
+        valuation: &Valuation<F>,
         alloc_ptrs: &mut HashMap<&'a String, AllocatedPtr<F>>,
         idx: usize,
     ) -> Result<()> {
-        if alloc_ptrs.contains_key(&self.input[idx]) {
-            bail!("{} already allocated", &self.input[idx]);
-        }
         let alloc_ptr = Self::allocate_ptr(
-            &mut cs.namespace(|| format!("allocate input {}", &self.input[idx])),
-            &store.hydrate_ptr(&witness.input[idx])?,
+            cs,
+            &store.hydrate_ptr(&valuation.input[idx])?,
             &self.input[idx],
+            alloc_ptrs,
         )?;
         alloc_ptrs.insert(&self.input[idx], alloc_ptr.clone());
         Self::inputize_ptr(cs, &alloc_ptr, &self.input[idx])?;
         Ok(())
     }
 
-    /// Create R1CS constraints for LEM given an evaluation witness.
+    fn on_concrete_path(concrete_path: &Boolean) -> Result<bool> {
+        concrete_path
+            .get_value()
+            .ok_or_else(|| anyhow!("Couldn't check whether we're on a concrete path"))
```

Comment:
Morally, I think these should all be `SynthesisError::AssignmentMissing`. Is there a strong reason not to use the `bellperson` errors for this — since they would do a better job of communicating exactly what the condition is and are standard practice?